File: src/main/java/gh/z0736190100/apparelstore/services/ApparelService.java
package gh.z0736190100.apparelstore.services;

import gh.z0736190100.apparelstore.models.ApparelDto;
import gh.z0736190100.apparelstore.models.ApparelPatchDto;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

/**
 * Service interface for Apparel operations
 */
public interface ApparelService {

    /**
     * Get all apparels
     * @return List of all apparels
     */
    List<ApparelDto> getAllApparels();

    /**
     * Get all apparels with optional filtering by apparel name and apparel style, with pagination
     * @param apparelName the apparel name to filter by (can be null)
     * @param apparelStyle the apparel style to filter by (can be null)
     * @param pageable pagination information
     * @return Page of apparels matching the criteria
     */
    Page<ApparelDto> getAllApparels(String apparelName, String apparelStyle, Pageable pageable);

    /**
     * Get a apparel by its ID
     * @param id the apparel ID
     * @return Optional containing the apparel if found
     */
    Optional<ApparelDto> getApparelById(Integer id);

    /**
     * Save a new apparel or update an existing one
     * @param apparelDto the apparel to save
     * @return the saved apparel
     */
    ApparelDto saveApparel(ApparelDto apparelDto);

    /**
     * Partially update an existing apparel
     * @param id the apparel ID
     * @param apparelPatchDto the apparel patch data
     * @return Optional containing the updated apparel if found
     */
    Optional<ApparelDto> patchApparel(Integer id, ApparelPatchDto apparelPatchDto);

    /**
     * Delete a apparel by its ID
     * @param id the apparel ID
     */
    void deleteApparelById(Integer id);
}

File: src/main/frontend/src/services/apparelService.ts
/**
 * Apparel Service
 *
 * Service for managing apparel-related operations
 */

import type { AxiosRequestConfig } from 'axios';
import apiService from './api';
import type { ApparelDto, ApparelPatchDto, PageOfApparelDto } from '../types/apparel';
import type { PaginationParams, SortParams, FilterParam } from '@utils/apiUtils';

// API endpoints
const BEER_API_URL = '/api/v1/apparels';
const BEER_BY_ID_URL = '/api/v1/apparels/{id}';

/**
 * Apparel Service class
 * Provides methods for interacting with the Apparel API
 */
class ApparelService {
  /**
   * Get a paginated list of apparels with optional filtering
   *
   * @param pagination - Pagination parameters
   * @param sort - Sorting parameters
   * @param apparelName - Optional apparel name filter
   * @param apparelStyle - Optional apparel style filter
   * @param config - Additional Axios request configuration
   * @returns Promise with the paginated apparel list
   */
  public async getApparels(
    pagination?: PaginationParams,
    sort?: SortParams,
    apparelName?: string,
    apparelStyle?: string,
    config?: AxiosRequestConfig
  ): Promise<PageOfApparelDto> {
    const filters: FilterParam[] = [];

    if (apparelName) {
      filters.push({ field: 'apparelName', value: apparelName });
    }

    if (apparelStyle) {
      filters.push({ field: 'apparelStyle', value: apparelStyle });
    }

    return apiService.getPaginatedWithNotification<PageOfApparelDto>(
      BEER_API_URL,
      pagination,
      sort,
      filters,
      config
    );
  }

  /**
   * Get a apparel by ID
   *
   * @param id - Apparel ID
   * @param config - Additional Axios request configuration
   * @returns Promise with the apparel
   */
  public async getApparelById(id: number, config?: AxiosRequestConfig): Promise<ApparelDto> {
    return apiService.getByIdWithNotification<ApparelDto>(BEER_BY_ID_URL, id, config);
  }

  /**
   * Create a new apparel
   *
   * @param apparel - Apparel data
   * @param config - Additional Axios request configuration
   * @returns Promise with the created apparel
   */
  public async createApparel(apparel: ApparelDto, config?: AxiosRequestConfig): Promise<ApparelDto> {
    return apiService.createWithNotification<ApparelDto>(BEER_API_URL, apparel, config);
  }

  /**
   * Update a apparel
   *
   * @param id - Apparel ID
   * @param apparel - Updated apparel data
   * @param config - Additional Axios request configuration
   * @returns Promise with the updated apparel
   */
  public async updateApparel(
    id: number,
    apparel: ApparelDto,
    config?: AxiosRequestConfig
  ): Promise<ApparelDto> {
    return apiService.updateWithNotification<ApparelDto>(BEER_BY_ID_URL, id, apparel, config);
  }

  /**
   * Partially update a apparel
   *
   * @param id - Apparel ID
   * @param apparelPatch - Partial apparel data
   * @param config - Additional Axios request configuration
   * @returns Promise with the updated apparel
   */
  public async patchApparel(
    id: number,
    apparelPatch: ApparelPatchDto,
    config?: AxiosRequestConfig
  ): Promise<ApparelDto> {
    return apiService.partialUpdateWithNotification<ApparelDto>(BEER_BY_ID_URL, id, apparelPatch, config);
  }

  /**
   * Delete a apparel
   *
   * @param id - Apparel ID
   * @param config - Additional Axios request configuration
   * @returns Promise with the response
   */
  public async deleteApparel(id: number, config?: AxiosRequestConfig): Promise<void> {
    return apiService.deleteResourceWithNotification<void>(BEER_BY_ID_URL, id, config);
  }
}

// Create a singleton instance
const apparelService = new ApparelService();

export default apparelService;
